chainer/functions/pooling/spatial_pyramid_pooling_2d.py

```diff
@@ -45,21 +46,34 @@ def __init__(self, x_shape, pyramid_height, pooling_class):
                 self.split_inds.append(out_dim)
 
     def forward(self, x):
+        self.retain_inputs((0,))
         self.ys = []
         for pooler in self.poolers:
-            y_var = pooler(*x)
+            y_var = pooler.apply(x)[0]
             n, c, h, w = pooler.out_shape = y_var.shape
             self.ys.append(y_var.reshape((n, c * h * w, 1, 1)))
 
         return concat.Concat(axis=1).forward([y.data for y in self.ys])
 
-    def backward(self, x, gy):
-        xp = cuda.get_array_module(*x)
-        gx = xp.zeros_like(x[0])
-        gys = xp.split(gy[0], self.split_inds, axis=1)
+    def backward(self, indexes, gy):
+        x, = self.get_retained_inputs()
+        return SpatialPyramidPooling2DGrad(self).apply((x, gy[0]))
+
+
+class SpatialPyramidPooling2DGrad(function_node.FunctionNode):
```

Comment:
`backward` is not implemented. If you are not planning to support double backprop of SPP (I think it's reasonable for v3), it's better to make `SpatialPyramidPooling2D` a subclass of `Function` instead of `FunctionNode` and directly implement the backward in this class (it looks `SpatialPyramidPooling2D` does not need to be a subclass of `Pooling2D`), then an error is raised when the user tries to differentiate SPP twice.

Reply:
Thank you for pointing out the lack of impl. of that. I added backward func. to `SpatialPyramidPooling2DGrad`.